corehq/apps/hqcase/api.py

```diff
@@ -0,0 +1,210 @@
+    case_name = jsonobject.StringProperty()
+    external_id = jsonobject.StringProperty()
+    user_id = jsonobject.StringProperty(required=True)
+    owner_id = jsonobject.StringProperty(name='@owner_id')
+    properties = jsonobject.DictProperty(validators=[is_simple_dict], default={})
+    indices = jsonobject.DictProperty(JsonIndex)
+    _is_case_creation = False
+
+    _allow_dynamic_properties = False
+
+    class Meta(object):
+        # prevent JsonObject from auto-converting dates etc.
+        string_conversions = ()
+
+    @memoized
+    def get_case_id(self):
+        return str(uuid.uuid4()) if self._is_case_creation else self.case_id
+
+    def get_caseblock(self):
+
+        def _if_specified(value):
+            return value if value is not None else CaseBlock.undefined
+
+        return CaseBlock(
+            case_id=self.get_case_id(),
+            user_id=self.user_id,
+            case_type=self.case_type if self._is_case_creation else CaseBlock.undefined,
+            case_name=_if_specified(self.case_name),
+            external_id=_if_specified(self.external_id),
+            owner_id=_if_specified(self.owner_id),
+            create=self._is_case_creation,
+            update=dict(self.properties),
+            index={
+                name: IndexAttrs(index.case_type, index.case_id, index.relationship)
+                for name, index in self.indices.items()
+            },
+        ).as_text()
+
+
+class JsonCaseCreation(BaseJsonCaseChange):
+    case_type = jsonobject.StringProperty(name='@case_type', required=True)
+    temporary_id = jsonobject.StringProperty()
+
+    # overriding from subclass to mark these required
+    case_name = jsonobject.StringProperty(required=True)
+    owner_id = jsonobject.StringProperty(name='@owner_id', required=True)
+
+    _is_case_creation = True
+
+
+class JsonCaseUpdate(BaseJsonCaseChange):
+    case_id = jsonobject.StringProperty(required=True)
+    _is_case_creation = False
+
+
+class UserError(Exception):
+    pass
+
+
+def handle_case_update(domain, data, user, device_id, case_id=None):
+    is_bulk = isinstance(data, list)
+    if is_bulk:
+        updates = _get_bulk_updates(domain, data, user)
+    else:
+        updates = [_get_individual_update(domain, data, user, case_id)]
+
+    xform, cases = _submit_case_updates(updates, domain, user, device_id)
+    if is_bulk:
+        return xform, cases
+    else:
+        return xform, cases[0] if cases else None
+
+
+def _get_individual_update(domain, data, user, case_id=None):
+    if case_id is not None and _missing_cases(domain, [case_id]):
+        raise UserError(f"No case found with ID '{case_id}'")
+
+    try:
+        return _get_case_update(data, user.user_id, case_id)
+    except BadValueError as e:
+        raise UserError(str(e))
+
+
+def _get_bulk_updates(domain, all_data, user):
+    if len(all_data) > 100:
+        raise UserError("You cannot submit more than 100 updates in a single request")
```

Comment:
100 is the number of cases per chunk we use in the case importer, so that seemed a reasonable enough number to use, given that we know HQ can handle that scale well enough.  We could chunk larger requests server-side, but since I imagine clients will be sending in batches anyways, pushing this constraint to the client seemed to be the most straightforward approach.  That way each request is transactional (you couldn't have partial failures even in form submission).

Reply:
I think it might be worth using the same constant for that then: 

https://github.com/dimagi/commcare-hq/blob/es%2Fcase_api/corehq/apps/case_importer/do_import.py#L35

Reply:
fair point, updated